packages/worker/src/worker.ts

```diff
@@ -1550,49 +1586,83 @@ export class Worker {
-   * This method would not wait for worker to complete shutdown if the inner `fnOrPromise` threw an error.
+   * **SDK versions `>=1.11.3`**:
+   * If the worker shuts down before the inner promise completes, allow no more than
+   * {@link RunUntilOptions.promiseCompletionTimeout} for the inner promise to complete,
+   * after which a {@link PromiseCompletionTimeoutError} is thrown.
    *
    * **SDK versions `>=1.5.0`**:
-   * This method always waits for both worker shutdown and inner `fnOrPromise` to resolve.
+   * This method always waits for both worker shutdown and inner `fnOrPromise` to complete.
    * If one of worker run -or- the inner promise throw an error, that error is rethrown.
    * If both throw an error, a {@link CombinedWorkerRunError} with a `cause` attribute containing both errors.
+   *
+   * **SDK versions `< 1.5.0`**:
+   * This method would not wait for worker to complete shutdown if the inner `fnOrPromise` threw an error.
+   *
+   * @returns the result of `fnOrPromise`
    */
-  async runUntil<R>(fnOrPromise: Promise<R> | (() => Promise<R>)): Promise<R> {
+  async runUntil<R>(fnOrPromise: Promise<R> | (() => Promise<R>), options?: RunUntilOptions): Promise<R> {
     const workerRunPromise = this.run();
+
+    let innerResult: PromiseSettledResult<R> | undefined;
     const innerPromise = (async () => {
       try {
-        const p = typeof fnOrPromise === 'function' ? fnOrPromise() : fnOrPromise;
-        return await p;
+        const result = await (typeof fnOrPromise === 'function' ? fnOrPromise() : fnOrPromise);
+        innerResult = { status: 'fulfilled', value: result };
+      } catch (err) {
+        innerResult = { status: 'rejected', reason: err };
       } finally {
         if (this.state === 'RUNNING') {
           this.shutdown();
         }
       }
     })();
-    const [innerResult, workerRunResult] = await Promise.allSettled([innerPromise, workerRunPromise]);
 
-    if (workerRunResult.status === 'rejected') {
-      if (innerResult.status === 'rejected') {
+    let workerError: Error | undefined;
+    try {
+      await workerRunPromise;
+    } catch (err) {
+      workerError = err as Error;
+    }
+
+    // Allow some extra time for the provided promise to resolve, if it hasn't already
+    const promiseCompletionTimeoutMs = msToNumber(options?.promiseCompletionTimeout ?? 1000);
+    if (innerResult === undefined && promiseCompletionTimeoutMs > 0) {
+      await timeoutPromise(innerPromise, promiseCompletionTimeoutMs);
+    }
+    if (innerResult === undefined)
+      innerResult = {
+        status: 'rejected',
+        reason: new PromiseCompletionTimeoutError(
+          `Promise did not resolve within ${promiseCompletionTimeoutMs}ms after Worker completed shutdown`
+        ),
+      };
+
+    if (workerError === undefined) {
+      if (innerResult.status === 'fulfilled') {
+        return innerResult.value;
+      } else {
+        throw innerResult.reason;
+      }
```

Comment:
Can `innerResult.reason` be missing? i.e., innerResult not undefined but the field is...

Reply:
It can't. The type mandates that `innerResult.reason` be defined if status is `rejected` (which is implied by `status !== 'fulfilled'`).